diff --git a/entity_usage.install b/entity_usage.install
index fb52e1a..a804d56 100644
--- a/entity_usage.install
+++ b/entity_usage.install
@@ -297,10 +297,20 @@ function entity_usage_update_8206(): void {
 
   // Drop all indices involving columns we want to change.
   $schema = $database->schema();
-  $schema->dropPrimaryKey('entity_usage');
   $schema->dropIndex('entity_usage', 'target_entity_string');
   $schema->dropIndex('entity_usage', 'source_entity_string');
 
+  // Add a temporary primary key column
+  $schema->addField('entity_usage', 'temp_id', [
+    'description' => 'Temporary ID for primary key workaround',
+    'type' => 'serial',
+    'unsigned' => TRUE,
+    'not null' => TRUE,
+  ]);
+
+  $schema->addPrimaryKey('entity_usage', ['temp_id']);
+
+  // Define the field specifications
   $specs = [
     'target_id_string' => [
       'description' => 'The target ID, when the entity uses string IDs.',
@@ -322,7 +332,10 @@ function entity_usage_update_8206(): void {
     $schema->changeField('entity_usage', $field, $field, $spec);
   }
 
-  // Add back the removed primary key index.
+  // Drop the temporary primary key
+  $schema->dropPrimaryKey('entity_usage');
+
+  // Add back the removed primary key index with the updated columns
   $schema->addPrimaryKey('entity_usage', [
     'target_id',
     'target_id_string',
@@ -335,6 +348,9 @@ function entity_usage_update_8206(): void {
     'field_name',
   ]);
 
+  // Remove the temporary primary key column.
+  $schema->dropField('entity_usage', 'temp_id');
+
   // Add back the removed indexes.
   $schema->addIndex('entity_usage', 'target_entity_string', [
     'target_type',
