diff --git a/modules/helfi_base_content/helfi_base_content.install b/modules/helfi_base_content/helfi_base_content.install
index a25784a7..f0c649ba 100644
--- a/modules/helfi_base_content/helfi_base_content.install
+++ b/modules/helfi_base_content/helfi_base_content.install
@@ -312,11 +312,11 @@ function helfi_base_content_update_9010() : void {
  * UHF-9471 revoke the admin user's permission for adding admin role to users.
  */
 function helfi_base_content_update_9011(): void {
-  if (!$role = Role::load('admin')) {
-    return;
-  }
-  $role->revokePermission('assign admin role');
-  $role->save();
+//  if (!$role = Role::load('admin')) {
+//    return;
+//  }
+//  $role->revokePermission('assign admin role');
+//  $role->save();
 }
 
 /**
